embassy-nrf/src/low_power_uarte.rs

```diff
@@ -0,0 +1,302 @@
+    /// A common pattern is to use [`stop()`](ReceiveFuture::stop) to cancel
+    /// unfinished transfers after a timeout to prevent lockup when no more data
+    /// is incoming.
+    ///
+    /// `rx_buffer` is marked as static as per `embedded-dma` requirements.
+    /// It it safe to use a buffer with a non static lifetime if memory is not
+    /// reused until the future has finished.
+    pub fn receive<'a, B>(&'a mut self, rx_buffer: B) -> ReceiveFuture<'a, B>
+    where
+        B: StaticWriteBuffer<Word = u8>,
+    {
+        // Panic if RX is running which can happen if the user has called
+        // `mem::forget()` on a previous future after polling it once.
+        assert!(!self.rx_started());
+
+        self.enable();
+
+        ReceiveFuture {
+            uarte: self,
+            buf: Some(rx_buffer),
+        }
+    }
+
+    fn rx_started(&self) -> bool {
+        self.instance.events_rxstarted.read().bits() != 0
+    }
+}
+
+/// Future for the [`LowPowerUarte::send()`] method.
+pub struct SendFuture<'a, B> {
+    uarte: &'a LowPowerUarte,
+    buf: B,
+}
+
+impl<'a, B> Drop for SendFuture<'a, B> {
+    fn drop(self: &mut Self) {
+        if self.uarte.tx_started() {
+            trace!("stoptx");
+
+            // Stop the transmitter to minimize the current consumption.
+            self.uarte
+                .instance
+                .tasks_stoptx
+                .write(|w| unsafe { w.bits(1) });
+            self.uarte.instance.events_txstarted.reset();
+        }
+    }
+}
+
+impl<'a, B> Future for SendFuture<'a, B>
+where
+    B: StaticReadBuffer<Word = u8>,
+{
+    type Output = ();
+
+    fn poll(self: core::pin::Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<()> {
+        if !self.uarte.tx_started() {
+            let uarte = &self.uarte.instance;
+
+            STATE.tx_done.reset();
+
+            let (ptr, len) = unsafe { self.buf.read_buffer() };
+            assert!(len <= EASY_DMA_SIZE);
+            // TODO: panic if buffer is not in SRAM
+
+            compiler_fence(Ordering::SeqCst);
+            uarte.txd.ptr.write(|w| unsafe { w.ptr().bits(ptr as u32) });
+            uarte
+                .txd
+                .maxcnt
+                .write(|w| unsafe { w.maxcnt().bits(len as _) });
+
+            trace!("starttx");
+            uarte.events_txstopped.reset();
+            uarte.tasks_starttx.write(|w| unsafe { w.bits(1) });
+        }
+
+        STATE.tx_done.poll_wait(cx)
+    }
+}
+
+/// Future for the [`LowPowerUarte::receive()`] method.
+pub struct ReceiveFuture<'a, B> {
+    uarte: &'a LowPowerUarte,
+    buf: Option<B>,
+}
+
+impl<'a, B> Drop for ReceiveFuture<'a, B> {
+    fn drop(self: &mut Self) {
+        if self.uarte.rx_started() {
+            trace!("stoprx");
+
+            self.uarte
+                .instance
+                .tasks_stoprx
+                .write(|w| unsafe { w.bits(1) });
+            self.uarte.instance.events_rxstarted.reset();
```

Comment:
I think it's necessary to do a blocking wait for ENDRX here. The DMA hardware may still write to RAM after triggering `tasks_stoprx`.

Reply:
That one is a good catch!
In most cases ENDRX has occurred already because the future will be dropped after the transfer has finished. But we abort the transfer we run into the exact problem you stated.

**Case 1: Drop after receive has finished**  
ENDRX has already been cleared by the ISR and will not be set again. But no DMA is ongoing -> Safe to drop

**Case 2: Drop the future during ongoing reception**  
STOPTX aborts the reception and the DMA transfer is still ongoing. We need to wait for ENDRX to be set before we safely drop the future.

The bust solution I could come up so far is to extend `Signal` with some `is_pending()` method and use that like: `while STATE.rx_done.is_pending() {}`
For Case 1 it is not pending and we drop immediately.
For Case 2 it blocks until the ISR triggers the signal

Reply:
sounds good! Perhaps a `signal.blocking_wait()` method would be better instead of `signal.is_pending()`.